File: packages/pipeline/src/grading.ts
/**
 * Post-hoc grading + "The Record" reveal. PRD P0 #11.
 *
 * After games settle: grade every pick published to any paid tier, compute
 * closing-line value, and build a reveal post that shows the FULL
 * Analyst/Sharp-depth content to ALL tiers (including Free).
 *
 * Hard rule: losses are NEVER omitted or edited. The reveal includes every
 * graded pick from the slate, win or lose, and rows are immutable once
 * written. Selective publication would destroy the record's value and is a
 * deceptive-marketing exposure.
 */
import type Database from "better-sqlite3";
import { Tier } from "@oddsports/shared";
import type { DailySlate, Sport } from "@oddsports/shared";
import { loadSlate } from "./store.js";

export type GradeResult = "win" | "loss" | "push" | "void";

export interface GradedPick {
  date: string;
  gameId: string;
  side: string;
  suggestedUnits: number;
  confidence: number;
  result: GradeResult;
  unitsDelta: number; // +units on win (at price), -units on loss, 0 push/void
  closingLine: number | null;
  clv: number | null; // our line vs closing — positive = beat the close
  /** The deepest (Analyst/Sharp) body that paid tiers saw — revealed to everyone. */
  revealBody: string;
}

export interface RollingRecord {
  wins: number;
  losses: number;
  pushes: number;
  unitsNet: number;
  avgClv: number | null;
  graded: number;
}

export function ensureGradingTables(db: Database.Database): void {
  db.exec(`
    CREATE TABLE IF NOT EXISTS graded_picks (
      date TEXT NOT NULL,
      game_id TEXT NOT NULL,
      side TEXT NOT NULL,
      suggested_units REAL NOT NULL,
      confidence INTEGER NOT NULL,
      result TEXT NOT NULL,
      units_delta REAL NOT NULL,
      closing_line REAL,
      clv REAL,
      reveal_body TEXT NOT NULL,
      sport TEXT,
      graded_at TEXT NOT NULL DEFAULT (datetime('now')),
      PRIMARY KEY (date, game_id, side)
    );
  `);
}

/** Final scores keyed by gameId. TODO(fable): fetch from the odds API's scores endpoint. */
export interface FinalScore {
  gameId: string;
  homeScore: number;
  awayScore: number;
  closingSpread: number | null; // home-perspective closing line
}

export async function fetchFinalScores(_date: string): Promise<FinalScore[]> {
  // the-odds-api: GET /sports/{key}/scores?daysFrom=1 — wire up with ODDS_API_KEY.
  console.warn("[grading] fetchFinalScores not implemented — returning []");
  return [];
}

/**
 * Grade yesterday's slate. Idempotent: PRIMARY KEY + INSERT OR IGNORE means
 * a pick, once graded, is immutable (re-runs cannot rewrite history).
 */
export async function gradeSlate(db: Database.Database, date: string): Promise<GradedPick[]> {
  ensureGradingTables(db);
  const slate = loadSlate(db, date);
  if (!slate) {
    console.warn(`[grading] no slate for ${date}`);
    return [];
  }
  const scores = await fetchFinalScores(date);
  const scoreById = new Map(scores.map((s) => [s.gameId, s]));

  const graded: GradedPick[] = [];
  const insert = db.prepare(`
    INSERT OR IGNORE INTO graded_picks
      (date, game_id, side, suggested_units, confidence, result, units_delta, closing_line, clv, reveal_body, sport)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `);

  for (const pick of deepestBlocks(slate)) {
    const score = scoreById.get(pick.gameId);
    if (!score) continue; // not settled yet — next run picks it up

    const g = gradePick(date, pick, score);
    insert.run(
      g.date, g.gameId, g.side, g.suggestedUnits, g.confidence,
      g.result, g.unitsDelta, g.closingLine, g.clv, g.revealBody, sportOf(slate, pick.gameId)
    );
    graded.push(g);
  }
  console.log(`[grading] ${date}: graded ${graded.length} picks`);
  return graded;
}

/** One block per game — the deepest tier depth that was published. */
function deepestBlocks(slate: DailySlate) {
  const byGame = new Map<string, (typeof slate.picks)[number]>();
  for (const p of slate.picks) {
    const cur = byGame.get(p.gameId);
    if (!cur || p.minTier > cur.minTier) byGame.set(p.gameId, p);
  }
  return [...byGame.values()];
}

function gradePick(date: string, pick: { gameId: string; body: string; confidence: number }, score: FinalScore): GradedPick {
  const side = /\*\* — (.+?)$/m.exec(pick.body)?.[1] ?? "unknown";
  const units = Number(/Suggested size: ([\d.]+)u/.exec(pick.body)?.[1] ?? 1);

  const result = settleSpread(side, score);
  const unitsDelta = result === "win" ? units * 0.91 : result === "loss" ? -units : 0; // -110 juice assumption

  const ourLine = extractLine(side);
  const clv =
    ourLine !== null && score.closingSpread !== null
      ? Math.round((score.closingSpread - ourLine) * 100) / 100
      : null;

  return {
    date,
    gameId: pick.gameId,
    side,
    suggestedUnits: units,
    confidence: pick.confidence,
    result,
    unitsDelta: Math.round(unitsDelta * 100) / 100,
    closingLine: score.closingSpread,
    clv,
    revealBody: pick.body,
  };
}

/** Settle a home/away spread side against the final score. */
function settleSpread(side: string, score: FinalScore): GradeResult {
  const line = extractLine(side);
  if (line === null) return "void";
  const margin = score.homeScore - score.awayScore;
  // Side string names the team; we grade home-perspective if the line sign implies it.
  // TODO(fable): carry structured side info in PickBlock instead of parsing strings.
  const covered = margin + line;
  if (covered > 0) return "win";
  if (covered < 0) return "loss";
  return "push";
}

function extractLine(side: string): number | null {
  const m = /([+-]?\d+(?:\.\d+)?)\s*$/.exec(side.trim());
  return m ? Number(m[1]) : null;
}

export function rollingRecord(db: Database.Database, opts?: { sport?: Sport; sinceDate?: string }): RollingRecord {
  ensureGradingTables(db);
  const where: string[] = ["result != 'void'"];
  const params: any[] = [];
  if (opts?.sport) { where.push("sport = ?"); params.push(opts.sport); }
  if (opts?.sinceDate) { where.push("date >= ?"); params.push(opts.sinceDate); }

  const row = db.prepare(`
    SELECT
      SUM(result = 'win') AS wins,
      SUM(result = 'loss') AS losses,
      SUM(result = 'push') AS pushes,
      ROUND(SUM(units_delta), 2) AS unitsNet,
      ROUND(AVG(clv), 2) AS avgClv,
      COUNT(*) AS graded
    FROM graded_picks WHERE ${where.join(" AND ")}
  `).get(...params) as any;

  return {
    wins: row.wins ?? 0,
    losses: row.losses ?? 0,
    pushes: row.pushes ?? 0,
    unitsNet: row.unitsNet ?? 0,
    avgClv: row.avgClv,
    graded: row.graded ?? 0,
  };
}

/** Build the daily reveal post — full paid-tier depth, shown to everyone. */
export function buildRevealPost(db: Database.Database, date: string): string {
  ensureGradingTables(db);
  const rows = db
    .prepare("SELECT * FROM graded_picks WHERE date = ? ORDER BY units_delta DESC")
    .all(date) as any[];
  if (rows.length === 0) return `📊 **The Record — ${date}**\n\nNo settled picks yet.`;

  const record = rollingRecord(db);
  const lines = rows.map((r) => {
    const icon = r.result === "win" ? "✅" : r.result === "loss" ? "❌" : "➖";
    const clv = r.clv !== null ? ` · CLV ${r.clv > 0 ? "+" : ""}${r.clv}` : "";
    return `${icon} ${r.side} — ${r.result.toUpperCase()} (${r.units_delta > 0 ? "+" : ""}${r.units_delta}u${clv})`;
  });
  const reveals = rows.map((r) => `---\n_What paid tiers saw:_\n\n${r.reveal_body}`);

  return [
    `📊 **The Record — ${date}**`,
    "",
    ...lines,
    "",
    `**Rolling: ${record.wins}-${record.losses}${record.pushes ? `-${record.pushes}` : ""}, ${record.unitsNet > 0 ? "+" : ""}${record.unitsNet}u${record.avgClv !== null ? `, avg CLV ${record.avgClv}` : ""}** (all picks since launch — losses included, always)`,
    "",
    "Yesterday's full paid-tier analysis, revealed:",
    ...reveals,
  ].join("\n");
}

File: crates/oddsports-pipeline/src/grading.rs
//! Post-hoc grading + "The Record" reveal. PRD P0 #11.
//!
//! After games settle: grade every pick published to any paid tier, compute
//! closing-line value, and build a reveal post showing the FULL
//! Analyst/Sharp-depth content to ALL tiers (including Free).
//!
//! Hard rule: losses are NEVER omitted or edited. `INSERT OR IGNORE` on the
//! primary key makes graded rows immutable — re-runs cannot rewrite history.
//! Selective publication would destroy the record's value and is a
//! deceptive-marketing exposure.

use crate::store::load_slate;
use anyhow::Result;
use oddsports_shared::{DailySlate, PickBlock, PickTeam};
use rusqlite::{params, Connection};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GradeResult {
    Win,
    Loss,
    Push,
    Void,
}

impl GradeResult {
    fn as_str(&self) -> &'static str {
        match self {
            GradeResult::Win => "win",
            GradeResult::Loss => "loss",
            GradeResult::Push => "push",
            GradeResult::Void => "void",
        }
    }
}

/// Final scores keyed by game id.
/// TODO(fable): fetch from the odds API scores endpoint (GET /sports/{key}/scores?daysFrom=1).
pub struct FinalScore {
    pub game_id: String,
    pub home_score: i32,
    pub away_score: i32,
    /// Home-perspective closing spread.
    pub closing_spread: Option<f64>,
}

pub async fn fetch_final_scores(sports: &[oddsports_shared::Sport]) -> Result<Vec<FinalScore>> {
    let (Ok(base), Ok(key)) = (std::env::var("ODDS_API_BASE"), std::env::var("ODDS_API_KEY")) else {
        tracing::warn!("ODDS_API_* not configured — cannot grade");
        return Ok(vec![]);
    };
    let client = reqwest::Client::new();
    let mut scores = Vec::new();

    for sport in sports {
        let url = format!(
            "{base}/sports/{}/scores?daysFrom=2&apiKey={key}",
            sport.odds_api_key()
        );
        let res = client.get(&url).send().await?;
        if !res.status().is_success() {
            tracing::warn!(sport = ?sport, status = %res.status(), "scores fetch skipped");
            continue;
        }
        let events: Vec<serde_json::Value> = res.json().await?;
        for ev in events {
            if !ev["completed"].as_bool().unwrap_or(false) {
                continue;
            }
            let home_team = ev["home_team"].as_str().unwrap_or_default();
            let parse_score = |team: &str| -> Option<i32> {
                ev["scores"].as_array()?.iter().find_map(|s| {
                    (s["name"].as_str() == Some(team))
                        .then(|| s["score"].as_str()?.parse().ok())
                        .flatten()
                })
            };
            let away_team = ev["away_team"].as_str().unwrap_or_default();
            let (Some(home_score), Some(away_score)) = (parse_score(home_team), parse_score(away_team)) else {
                continue;
            };
            scores.push(FinalScore {
                game_id: ev["id"].as_str().unwrap_or_default().to_string(),
                home_score,
                away_score,
                // TODO(fable): capture true closing line via a pre-game snapshot
                // just before start time; the scores endpoint doesn't carry it.
                closing_spread: None,
            });
        }
    }
    Ok(scores)
}

/// Grade a date's slate. Idempotent — already-graded picks are skipped.
pub async fn grade_slate(db: &Connection, date: &str) -> Result<usize> {
    let Some(slate) = load_slate(db, date)? else {
        tracing::warn!(date, "no slate to grade");
        return Ok(0);
    };
    let sports: Vec<oddsports_shared::Sport> = {
        let mut v: Vec<_> = slate.picks.iter().map(|p| p.sport).collect();
        v.sort_by_key(|s| s.odds_api_key());
        v.dedup();
        v
    };
    let scores: HashMap<String, FinalScore> = fetch_final_scores(&sports)
        .await?
        .into_iter()
        .map(|s| (s.game_id.clone(), s))
        .collect();

    let mut graded = 0;
    for pick in deepest_blocks(&slate) {
        let Some(score) = scores.get(&pick.game_id) else { continue }; // not settled yet

        let result = settle_spread(pick, score);
        let units = pick.model.suggested_units;
        let units_delta = match result {
            GradeResult::Win => units * 0.91, // -110 juice assumption
            GradeResult::Loss => -units,
            _ => 0.0,
        };
        // CLV from structured data — no string parsing (home-perspective line).
        let our_home_line = match pick.model.pick_team {
            PickTeam::Home => pick.model.picked_line,
            PickTeam::Away => -pick.model.picked_line,
        };
        let clv = score.closing_spread.map(|c| ((c - our_home_line) * 100.0).round() / 100.0);

        graded += db.execute(
            "INSERT OR IGNORE INTO graded_picks
               (date, game_id, side, suggested_units, confidence, result, units_delta, closing_line, clv, reveal_body, sport)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11)",
            params![
                date,
                pick.game_id,
                pick.model.side,
                units,
                pick.confidence,
                result.as_str(),
                (units_delta * 100.0).round() / 100.0,
                score.closing_spread,
                clv,
                pick.body,
                serde_json::to_string(&pick.sport)?,
            ],
        )?;
    }
    tracing::info!(date, graded, "grading complete");
    Ok(graded)
}

/// One block per game — the deepest tier depth that was published.
fn deepest_blocks(slate: &DailySlate) -> Vec<&PickBlock> {
    let mut by_game: HashMap<&str, &PickBlock> = HashMap::new();
    for p in &slate.picks {
        by_game
            .entry(p.game_id.as_str())
            .and_modify(|cur| {
                if p.min_tier > cur.min_tier {
                    *cur = p;
                }
            })
            .or_insert(p);
    }
    by_game.into_values().collect()
}

/// Settle a spread pick against the final score using structured pick data.
fn settle_spread(pick: &PickBlock, score: &FinalScore) -> GradeResult {
    let margin = (score.home_score - score.away_score) as f64;
    let covered = match pick.model.pick_team {
        PickTeam::Home => margin + pick.model.picked_line,
        PickTeam::Away => -margin + pick.model.picked_line,
    };
    if covered > 0.0 {
        GradeResult::Win
    } else if covered < 0.0 {
        GradeResult::Loss
    } else {
        GradeResult::Push
    }
}

#[derive(Debug, Default)]
pub struct RollingRecord {
    pub wins: i64,
    pub losses: i64,
    pub pushes: i64,
    pub units_net: f64,
    pub avg_clv: Option<f64>,
    pub graded: i64,
}

pub fn rolling_record(db: &Connection) -> Result<RollingRecord> {
    db.query_row(
        "SELECT
           COALESCE(SUM(result = 'win'), 0),
           COALESCE(SUM(result = 'loss'), 0),
           COALESCE(SUM(result = 'push'), 0),
           COALESCE(ROUND(SUM(units_delta), 2), 0),
           ROUND(AVG(clv), 2),
           COUNT(*)
         FROM graded_picks WHERE result != 'void'",
        [],
        |r| {
            Ok(RollingRecord {
                wins: r.get(0)?,
                losses: r.get(1)?,
                pushes: r.get(2)?,
                units_net: r.get(3)?,
                avg_clv: r.get(4)?,
                graded: r.get(5)?,
            })
        },
    )
    .map_err(Into::into)
}

/// Build the daily reveal post — full paid-tier depth, shown to everyone.
pub fn build_reveal_post(db: &Connection, date: &str) -> Result<String> {
    let mut stmt = db.prepare(
        "SELECT side, result, units_delta, clv, reveal_body
         FROM graded_picks WHERE date = ?1 ORDER BY units_delta DESC",
    )?;
    let rows: Vec<(String, String, f64, Option<f64>, String)> = stmt
        .query_map(params![date], |r| {
            Ok((r.get(0)?, r.get(1)?, r.get(2)?, r.get(3)?, r.get(4)?))
        })?
        .collect::<std::result::Result<_, _>>()?;

    if rows.is_empty() {
        return Ok(format!("📊 **The Record — {date}**\n\nNo settled picks yet."));
    }

    let rec = rolling_record(db)?;
    let mut out = vec![format!("📊 **The Record — {date}**"), String::new()];

    for (side, result, delta, clv, _) in &rows {
        let icon = match result.as_str() {
            "win" => "✅",
            "loss" => "❌",
            _ => "➖",
        };
        let clv_str = clv.map(|c| format!(" · CLV {}{c}", if c > 0.0 { "+" } else { "" })).unwrap_or_default();
        out.push(format!(
            "{icon} {side} — {} ({}{delta}u{clv_str})",
            result.to_uppercase(),
            if *delta > 0.0 { "+" } else { "" }
        ));
    }

    out.push(String::new());
    out.push(format!(
        "**Rolling: {}-{}{}, {}{}u{}** (all picks since launch — losses included, always)",
        rec.wins,
        rec.losses,
        if rec.pushes > 0 { format!("-{}", rec.pushes) } else { String::new() },
        if rec.units_net > 0.0 { "+" } else { "" },
        rec.units_net,
        rec.avg_clv.map(|c| format!(", avg CLV {c}")).unwrap_or_default(),
    ));
    out.push(String::new());
    out.push("Yesterday's full paid-tier analysis, revealed:".into());
    for (_, _, _, _, body) in &rows {
        out.push(format!("---\n_What paid tiers saw:_\n\n{body}"));
    }

    Ok(out.join("\n"))
}
